### PR TITLE
Post api

### PR DESCRIPTION
**Pull request by humulla**
A post endpoint has been added to allow the JSON file to save to the dataDropBucket S3 Bucket, relevant test has also been added.

Would you please be able to check if this has been implemented properly and let me know if they are any further improvements for me to implement.

### CODE REVIEW

File: ElectoralPOC/V1/Infrastructure/AwsS3Client.cs [hunk 1]
```diff
@@ -3,31 +3,33 @@
...
+                byte[] _byteArray = Encoding.ASCII.GetBytes("unprocessed /{ }_requestdata.json");
+                using (var stream = new MemoryStream(_byteArray))
                 {
                     var objectRequest = new PutObjectRequest
                     {
-                        BucketName = "electoral-register-stack-rdatadropbucketa4dd7fb3-jcd1deylvkwn",
-                        Key = "unprocessed /{ }_requestdata.json",
-                        ContentType = "application/json",
+                        BucketName = Environment.GetEnvironmentVariable("electoral-register-stack-rdatadropbucketa4dd7fb3-jcd1deylvkwn"),
+                        Key = Environment.GetEnvironmentVariable("unprocessed /{ }_requestdata.json"),
```
Comment (MirelaGeorgiou) [inline ElectoralPOC/V1/Infrastructure/AwsS3Client.cs]: The 'Key' value should come from whatever was passed as in the POST request. Ideally, you would first use a helper method in the use case to ensure that the file name supplied in the request has '.json' in the file name to define the correct file extension

File: ElectoralPOC/V1/Infrastructure/AwsS3Client.cs [hunk 2]
```diff
@@ -3,31 +3,33 @@
...
+              
+                byte[] _byteArray = Encoding.ASCII.GetBytes("unprocessed /{ }_requestdata.json");
+                using (var stream = new MemoryStream(_byteArray))
                 {
                     var objectRequest = new PutObjectRequest
                     {
-                        BucketName = "electoral-register-stack-rdatadropbucketa4dd7fb3-jcd1deylvkwn",
-                        Key = "unprocessed /{ }_requestdata.json",
-                        ContentType = "application/json",
+                        BucketName = Environment.GetEnvironmentVariable("electoral-register-stack-rdatadropbucketa4dd7fb3-jcd1deylvkwn"),
```
Comment (MirelaGeorgiou) [inline ElectoralPOC/V1/Infrastructure/AwsS3Client.cs]: Please change the name of the env var to something different or if we want to make this API more generic, you could add 'BucketName' to the POST request object and pass it here.

File: ElectoralPOC/V1/Infrastructure/AwsS3Client.cs [hunk 3]
```diff
@@ -3,31 +3,33 @@
...
-
-        public string GenerateS3PreSignURL(GetPreSignedUrlRequest request)
+        public string SaveJsonToS3(GetPreSignedUrlRequest request)
         {
             using (AmazonS3Client _s3Client = new AmazonS3Client(RegionEndpoint.USWest2))
             {
-
-                using (var stream = new MemoryStream())
+              
+                byte[] _byteArray = Encoding.ASCII.GetBytes("unprocessed /{ }_requestdata.json");
```
Comment (MirelaGeorgiou) [inline ElectoralPOC/V1/Infrastructure/AwsS3Client.cs]: This should be retrieving the bytes from the JsonData passed in the POST request

File: ElectoralPOC/V1/Gateways/SaveJsonToS3Gateway.cs [hunk 4]
```diff
@@ -30,19 +31,15 @@ public string GenerateS3PutPreSignurl(GenerateS3PreSignedUrlRequest request)
...
-            }
-            catch (UrlExpirationTimeInvalidException)
-            {
-                throw;
+                return _awsS3Client.SaveJsonToS3(getUrlRequest);
             }
             catch (Exception ex) //TODO find list of specific AWS exceptions that could be thrown by S3
             {
-                throw new PreSignedUrlCouldNotBeGeneratedException($"Pre-signed URL could not be generated - {ex.Message}, {ex.InnerException}");
+                throw new SaveJsonToS3CouldNotBeGeneratedException($"Pre-signed URL could not be generated - {ex.Message}, {ex.InnerException}");
```
Comment (MirelaGeorgiou) [inline ElectoralPOC/V1/Gateways/SaveJsonToS3Gateway.cs]: Exception message needs to be changed

File: ElectoralPOC/V1/Gateways/SaveJsonToS3Gateway.cs [hunk 5]
```diff
@@ -30,19 +31,15 @@ public string GenerateS3PutPreSignurl(GenerateS3PreSignedUrlRequest request)
                 GetPreSignedUrlRequest getUrlRequest = new GetPreSignedUrlRequest
```
Comment (MirelaGeorgiou) [inline ElectoralPOC/V1/Gateways/SaveJsonToS3Gateway.cs]: This request object will no longer be needed

File: ElectoralPOC/V1/Domain/Exceptions/SaveJsonToS3CouldNotBeGeneratedException.cs [hunk 6]
```diff
@@ -5,8 +5,8 @@
 
 namespace ElectoralPOC.V1.Domain.Exceptions
 {
-    public class PreSignedUrlCouldNotBeGeneratedException : Exception
+    public class SaveJsonToS3CouldNotBeGeneratedException : Exception
     {
-        public PreSignedUrlCouldNotBeGeneratedException(string message) : base(message) { }
+        public SaveJsonToS3CouldNotBeGeneratedException(string message) : base(message) { }
```
Comment (MirelaGeorgiou) [inline ElectoralPOC/V1/Domain/Exceptions/SaveJsonToS3CouldNotBeGeneratedException.cs]: Please change the exception name to something reflecting the functionality of this API endpoint (e.g. JsonFileCouldNotBeSavedToS3Exception)

File: ElectoralPOC/V1/Controllers/SaveJsonToS3Controller.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+using Microsoft.AspNetCore.Mvc;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ElectoralPOC.V1.Controllers
+{
+    [ApiController]
+    [Route("unprocessed /{ }_requestdata.json")]
```
Comment (MirelaGeorgiou) [inline ElectoralPOC/V1/Controllers/SaveJsonToS3Controller.cs]: The route needs to be updated (e.g api/v1/json-data

File: ElectoralPOC.Tests/V1/UseCase/SaveJsonToS3UseCaseTests.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ElectoralPOC.Tests.V1.UseCase
+{
+    public class SaveJsonToS3UseCaseTests
```
Comment (MirelaGeorgiou) [inline ElectoralPOC.Tests/V1/UseCase/SaveJsonToS3UseCaseTests.cs:17]: Usecase method names need to be updated (they still save GetS3PutPresignUrl)

File: ElectoralPOC.Tests/V1/Gateways/SaveJsonToS3GatewayTests.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        private SaveJsonToS3Gateway _classUnderTest;
+        private Mock<IAwsS3Client> _mockS3Client;
+        private readonly Fixture _fixture = new Fixture();
+        [SetUp]
+        public void Setup()
+        {
+            _mockS3Client = new Mock<IAwsS3Client>();
+            _classUnderTest = new SaveJsonToS3Gateway(_mockS3Client.Object);
+            Environment.SetEnvironmentVariable("PRESIGNED_URL_EXPIRATION_IN_SECONDS", "0.2");
```
Comment (MirelaGeorgiou) [inline ElectoralPOC.Tests/V1/Gateways/SaveJsonToS3GatewayTests.cs]: This env var will no longer be needed

File: ElectoralPOC.Tests/V1/Gateways/SaveJsonToS3GatewayTests.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        public void CheckThatIfS3ClientThrowsExceptionPreSignedUrlCouldNotBeGeneratedExceptionIsThrown()
+        {
+            _mockS3Client.Setup(x => x.SaveJsonToS3(It.IsAny<GetPreSignedUrlRequest>())).Throws(new AggregateException());
+            Assert.Throws<SaveJsonToS3CouldNotBeGeneratedException>(() => _classUnderTest.ConvertJsonToArray(new SaveJsonToS3Request()));
+        }
+
+        [Test]
+        public void CanSuccessfullyReturnGeneratedUrl()
+        {
+            var expectedResponse = "https://master.d1ew52s1hpob9x.amplifyapp.com/form/register-applicant/applicant-details";
```
Comment (MirelaGeorgiou) [inline ElectoralPOC.Tests/V1/Gateways/SaveJsonToS3GatewayTests.cs]: If we are not returning anything (just a 201 response), then our use case should not be returning anything either

**Review by MirelaGeorgiou — COMMENTED** [review]
Hi Humairaa, please review the comments :)

File: ElectoralPOC/V1/Infrastructure/AwsS3Client.cs (hunk 2 above)
Comment (humulla) [reply ElectoralPOC/V1/Infrastructure/AwsS3Client.cs]: I'm confused, BucketName is being called from the PutObjectRequest which is then being passed here.

File: ElectoralPOC.Tests/V1/Helper/SaveJsonToS3HelperTests.cs [hunk 11]
```diff
@@ -30,8 +32,17 @@ public void AnyStartingForwardSlashesWillBeRemovedIfPresent()
             var submissionId = _faker.Random.AlphaNumeric(9);
             var expectedUrl = submissionId + "/docs/" + fileName;
 
-            var actualResult = GenerateUrlHelper.ComposeFilePath("", "/"+fileName, "/"+submissionId);
+            var actualResult = SaveJsonToS3Helper.ComposeFilePath("", "/" + fileName, "/" + submissionId);
             actualResult.Should().Be(expectedUrl);
         }
+
+        [Test]
```
Comment (MirelaGeorgiou) [inline ElectoralPOC.Tests/V1/Helper/SaveJsonToS3HelperTests.cs:44]: Maybe add 1-2 more tests covering the 'negative' scenarios when an invalid extension is provided

**Review by MirelaGeorgiou — APPROVED** [review]
Looks good :) Please see my comment regarding adding more test coverage

File: ElectoralPOC.Tests/V1/Helper/SaveJsonToS3HelperTests.cs (hunk 11 above)
Comment (humulla) [reply ElectoralPOC.Tests/V1/Helper/SaveJsonToS3HelperTests.cs:39]: I did this test in the controller test section as I wasn't too sure how to write it over here

Comment (humulla) [reply ElectoralPOC.Tests/V1/Helper/SaveJsonToS3HelperTests.cs:44]: This test has been completed in the controller section, happy to move it over if you would like me too?
